text_quality function added

HarjyotKaur opened the pull request:
Added function text_quality for package PySyntext with docstings. 
Please review it and let me know you thoughts.

pak-alex commented:
Overall, I think the code is solid. The functions are written in a clear, readable manner, and most of the functionality for text_quality is in place. A few things that came up while bug testing, in order of importance: 

* Potentially large issue: sets cannot contain duplicate items. If there is one word spelled wrong multiple times (or one toxic word used multiple times), set() only counts one instance of it, messing up proportion calculations. (ie "sppelled sppelled wrong" would give proportion 1/3 when true is 2/3) 

* A filtering error occurs when every word in the input string is a Proper Noun (ie. "Alex Ian Aaron Jessie Paris" results in an error); I think it has something to do with the way Proper Nouns are treated / readded back into total word count after they are tagged? 

* Exceptions should be raised in the function to handle cases where input is not a string, and also to satisfy the verify_input1-3 functions in the test_text_quality.py

* The test_output_spell_error and test_output_toxicity functions in test_text_quality.py will fail, because the example proportions are wrong. Correct proportions should be around ~0.077 (1/13) for both spell error and toxicity proportion for the example sentence, however the example spell error proportion is 0.15 and the example toxicity proportion is 0.08. 

* Not too sure if important: weird edge cases where if every word in a sentence is spelled wrong, and the first word is capitalized, tagging thinks the first word is a proper noun; maybe handle the case where every word is spelled wrong separately? 

* Not too sure if important: Maybe add in a round function for the proportion of words spelled wrong / proportion of toxic words? 

* I believe that the text_quality.py script is importing stopwords from nltk.corpus without actually using it anywhere, but I could be wrong. 

There are also a couple documentation issues that may be worth fixing: 

* The returns section in spell_check / toxicity_check docstring could be tweaked; returns a dict of (list, int, float), rather than list, int, double

* Typo in text_quality parameters: input string "to" be analyzed

* Returns section in clean() function does not have correct output

* On a similar note to the 4'th bullet point in the bugs mentioned above: docstring examples for toxicity_check and spell_check seem to be inconsistent (ie. the example for toxicity_check() currently says "this is so shitty" results in a proportion of ~0.0588, when I believe the correct proportion is 0.25) 

* Need to update test_text_quality.py to include tests for the new columns in the dataframe (as well as update readme)

HarjyotKaur commented:
Hi Alex,
Thanks for reviewing the code. 
Please find below the above mentioned bugs addressed:

- Same word misspelled multiple times: Fixed
 (ie "sppelled sppelled wrong" would give proportion  2/3)

- Only nouns : Fixed (ie. "Alex Ian Aaron Jessie Paris" should not result in an error

- Exceptions should be raised in the function to handle cases where input is not a string, and also to satisfy the verify_input1-3 functions in the test_text_quality.py: The test cases are no more valid as the output dataframe has been changed. As far as verify_input1-3 are concerned, if a non string variable is entered such as 123 or @#$%^&&* it would raise an issue whereas if an empty string is entered "" or a string with only special characters which would be removed in the cleaning "!@#$%^&*" it should return a dataframe with no spelling mistakes and no toxicity content, which the function does. 

- The test_output_spell_error and test_output_toxicity functions in test_text_quality.py: Fixed example in test case. 

- A capitalized word spelt wrong: Potential bottlenecks: First word of a sentence spelt wrong ( rare case). This is a trade-off I am willing to take else all the nouns would be marked as spelling errors. The algorithm assumes that the text entered by the person follows basic rules of grammar i.e "The basic capitalization rule of proper nouns is that the first letters are capitalized." Also, if you do come across a dictionary of all possible proper nouns (singular and plural) please share it, that would be a fix for this. I would also discuss this with our TA, to see any possible solutions.

- A round function for proportions: I want to discuss this with the TA. (Scikit learn does not round the score of the model)

- Stop words are not being used, removed import statement. 

- Docstring and documentation typos fixed

- The return of the function has been changed during developement phase, I am yet to update the test cases for the same.

pak-alex commented:
Looks like issues were fixed. Merging!
